### PR TITLE
Upgrade to wayland-rs 0.21

### PR DESCRIPTION
**Pull request by trimental**
This PR upgrades the smithay-client-toolkit to use wayland-rs 0.21 (pure rust implementation :tada:)

This PR makes necessary breaking modifications to the API but I've tried to minimize the changes, the API can be changed to better fit wayland-rs 0.21 if you'd like. I've run this through clippy and tested the examples but I'll be porting winit to use this PR to test the changes in more depth as well as reading through the diff a bit more.

### CODE REVIEW

File: src/pointer/mod.rs [hunk 1]
```diff
@@ -55,25 +56,33 @@ impl AutoThemer {
...
+    pub fn theme_pointer_with_impl<Impl, UD>(
         &self,
-        pointer: NewProxy<wl_pointer::WlPointer>,
+        seat: &Proxy<wl_seat::WlSeat>,
         mut implementation: Impl,
+        user_data: UD,
     ) -> AutoPointer
     where
-        Impl: Implementation<AutoPointer, wl_pointer::Event> + Send + 'static,
+        Impl: FnMut(AutoPointer, wl_pointer::Event) + Send + 'static,
```
Comment (vberger) [inline src/pointer/mod.rs]: Same here for the order of arguments.

File: src/pointer/theme.rs [hunk 2]
```diff
@@ -32,7 +32,7 @@ impl ThemeManager {
     pub fn init(
         name: Option<&str>,
         compositor: Proxy<wl_compositor::WlCompositor>,
-        shm: Proxy<wl_shm::WlShm>,
+        shm: &Proxy<wl_shm::WlShm>,
```
Comment (vberger) [inline src/pointer/theme.rs]: Why this change?

File: src/utils/mempool.rs [hunk 3]
```diff
@@ -41,7 +40,7 @@ impl DoubleMemPool {
     /// Create a double memory pool
     pub fn new<Impl>(shm: &Proxy<wl_shm::WlShm>, implementation: Impl) -> io::Result<DoubleMemPool>
     where
-        Impl: Implementation<(), ()> + Send,
+        Impl: FnMut((), ()) + Send + 'static,
```
Comment (vberger) [inline src/utils/mempool.rs]: The `()` arguments can be removed here too.

File: src/window/basic_frame.rs [hunk 4]
```diff
@@ -228,7 +229,7 @@ impl Frame for BasicFrame {
         compositor: &Proxy<wl_compositor::WlCompositor>,
         subcompositor: &Proxy<wl_subcompositor::WlSubcompositor>,
         shm: &Proxy<wl_shm::WlShm>,
-        implementation: Box<Implementation<u32, FrameRequest> + Send>,
+        implementation: Box<FnMut(u32, FrameRequest) + Send>,
```
Comment (vberger) [inline src/window/basic_frame.rs]: I'd prefer the arguments of this closure swapped to. Main matter (the requests) first, additional data second.

File: src/window/mod.rs [hunk 5]
```diff
@@ -60,7 +59,7 @@ pub enum Event {
 struct WindowInner<F> {
     frame: Arc<Mutex<F>>,
     shell_surface: Arc<Box<shell::ShellSurface>>,
-    user_impl: Box<Implementation<(), Event> + Send>,
+    user_impl: Box<FnMut((), Event) + Send>,
```
Comment (vberger) [inline src/window/mod.rs]: The `()` arugment can be dropped here too.

File: src/data_device/data_device.rs [hunk 6]
```diff
@@ -267,7 +275,7 @@ impl DataDevice {
     ///
     /// Correspond to traditional copy/paste behavior. Setting the
     /// source to `None` will clear the selection.
-    pub fn set_selection(&self, source: Option<DataSource>, serial: u32) {
+    pub fn set_selection(&self, source: &Option<DataSource>, serial: u32) {
```
[the remark this answers is not in the document]
Comment (vberger) [reply src/data_device/data_device.rs]: I'd prefer to keep this PR focused on the upgrade, and leave the clippy improvements as an other PR, to handle these questions separately.

**Comment by trimental** [comment]
I'll go through the documentation where changes were made

File: src/env.rs [hunk 7]
```diff
@@ -74,50 +71,44 @@ impl Environment {
...
     ) -> io::Result<Environment>
     where
-        Impl: Implementation<Proxy<wl_registry::WlRegistry>, GlobalEvent> + Send + 'static,
+        Impl: FnMut(GlobalEvent, Proxy<wl_registry::WlRegistry>) + Send + 'static,
     {
         let outputs = ::output::OutputMgr::new();
         let outputs2 = outputs.clone();
 
-        let manager = GlobalManager::new_with_cb(registry, move |event, registry: Proxy<_>| {
+        let manager = GlobalManager::new_with_cb(&display, move |event, registry| {
```
Comment (vberger) [inline src/env.rs]: Actually, for this to be correct now that we create the registry ourselves, we need to ensure that the registry will be created on the event queue that was given as argument.

To ensure that, we need to first make a wrapper of the display:

```rust
let display_wrapper = display.make_wrapper(&evq.get_token());
```

and then give this wrapper to `GlobalManager::new_with_cb(..)`

**Comment by vberger** [comment]
Okay, there this one last thing, and with that I think it'll be good. Thanks! :+1:

**Comment by trimental** [comment]
Looks like rust fmt has changed in 1.29 but I'll leave the formatting for a separate 'Cargo fmt' commit
